### DISCUSSION TITLE
Use ESP32

### DISCUSSION THREAD

**Discussion by Kunstflieger** [MiczFlor/RPi-Jukebox-RFID#1287] [Ideas]
I´m wondering if an ESP32 could replace the RPI which is an oversized device for this task.
This would bring lot of benefits like no boottime, no shutdownprocess, less power consumption...

Is there somebody who investigated in this topic? I have to less knowledge about programming in this "high" level.

May this could help: https://github.com/feelfreelinux/cspot

**Comment by s-martin** [comment]
At the moment I suspect a ESP is not powerful enough, as users already experience performance issues when using a small Pi Zero at least with the Spotify edition.


But new ideas and improvements are always welcome.

**Comment by Kunstflieger** [comment]
Thats a constructive feedback, which make sense. Thanks for this feedback.
I come with this idea from [Wlanthermo](https://github.com/WLANThermo-nano/WLANThermo_ESP32_Software/releases) which was also on RPI before it was transfered to an ESP32.
Than we will wait until new powerful devices will come on the market.

**Reply by thiscantbeserious** [reply]
If you're into dedictated low power hardware then I recommend looking into this here:

https://github.com/toniebox-reverse-engineering/hackiebox_cfw

It's an attempt to reverse engineer a custom firmware for the original Toniebox - which this projects somehow orignated from (the box, not the project above).

The Toniebox itself is using a Texas Instruments TI CC3200 SoC - but the same limitations hold true. You are not able to run a fully fledged OS on this - and writing a custom dedecitated OS is not only a LOT harder but also still applies to certain limits.

That's also why the original Toniebox is really limited in terms of Sound Quality.

**Comment by elwoodoo** [comment]
It **is** possible to build a very versatile player or audio project platform using an esp32, and I've come pretty far already, including hardware (not just built by me, I'm not alone in this project). I'll be able to release it in some weeks time. It consists of audio playback via class D amp, a little keyboard controller, of course Bluetooth and WiFi capability, a rotary encoder, buttons, and an i2c bus to be open for more hardware (like of course an rfid reader, a clock or sensors). Without all the extras, power consumption is pretty impressive low, because esp32 can do work in short bursts with its 2 cores at 240MHz.
I did not start by copying this project, though. Going with an embedded processor and an RTOS kernel needs some very different approaches. Now I wonder if it's worth it to establish some kind of compatibility between open source players? What does overlap naturally, and how much effort may it be worth to put into any level of compatibility? I'm not sure. Possibly in the end it's easier to convert config or content files between players _outside the players_  and without bothering to put that feature into the firmware. We'll see.

**Reply by pabera** [reply]
There are always pros and cons with each approach: micro controller vs embedded systems like a Pi. There is no doubt a ESP32 or Arduino are far better in energy consumption and the physical sizes make small devices possible. Their greatest disadvantage over a Pi is the specific development environment. C/C++ may be a lot harder to learn, develop and test compared to Python, Shell or JavaScript which can limit contributions from others. One of the advantages of why this project is very successful is that everyone can easily contribute to it at any level of knowledge.
